### ISSUE TITLE
Keep PO headers unchanged for existing catalogs

### ISSUE THREAD

**Issue by FredyC** [lingui/js-lingui#288]
**Describe the bug**
When I load extracted PO catalogs into some sort of editor and do some changes, it does modify headers in its own way. Later when I rerun `extract` again, headers are again modified, mostly lower casing, but nothing is lost.

This is mostly an issue about file versioning as when headers will be changing between commits, it means a bunch of unrelated changes all the time.

**To Reproduce**
You can use for example known POEdit to see this happening or simply edit PO file manually to change the casing of some header.

**Expected behavior**
Headers should be kept completely unchanged for existing catalogs.

**Additional context**
This is upstream issue https://github.com/smhg/gettext-parser/issues/12 which does not seem to be getting fixed any time soon. So it's either about forking it or copying out relevant parts.

**Comment by tricoder42** [comment]
`gettext-parser` also doesn't handle obsolete packages in a way I want, so maybe we should write own parser. Before I tried to find alternatives for `gettext-parser`, but nothing seemed to be better.

What about write the parser using [peg.js](https://pegjs.org/online). We need just read/write po file, nothing else.

**Comment by FredyC** [comment]
I am not feeling for PegJS, it says simple, but my head is spinning just by looking at examples :)

I have also found this one, so it's probably worth the try before getting hammer :) https://github.com/rubenv/pofile

**Comment by tricoder42** [comment]
Looks good! Interested in trying it on your own? :)

**Comment by FredyC** [comment]
Can you please elaborate more on how _obsolete_ should be working differently? It's just a flag that should go after `#,` right? I wasn't investigating much what's wrong with the current implementation.

Strangely enough that `pofile` package tries to identify obsolete items with prefix `#~` which doesn't seem to be anything official. It does parse flags per references just fine, but this is something extra.

**Comment by tricoder42** [comment]
I'm not sure if there's official specification for obsolete messages. I used to work with Django web framework and they simply comment out obsolete messages. I use `obsolete` flag as a workaround, because `gettext-parser` doesn't allow commenting messages.

**Comment by FredyC** [comment]
Alright, I see now. The `#~` seems to be an actual comment in PO file ([see example](https://github.com/rubenv/pofile/blob/master/test/fixtures/commented.po)). Based on this the module decides that particular translation is obsolete. I will set `obsolete` flag when message is commented out, seems like best course of action.

There is also support for `msgctxt` which denotes context. Based on reference doc, when the context is present, you can have a translation with the same `msgid` but with different context. That sounds like very useful feature in cases you are using messages as keys and sometimes it might be unclear in which context that message has been used. I have verified it with Loco service and it indeed works, some other will probably support it too. I will try to implement it from the side of format and perhaps there could a special macro that would add context to the message.

**Comment by tricoder42** [comment]
> Alright, I see now. The #~ seems to be an actual comment in PO file (see example). Based on this the module decides that particular translation is obsolete. I will set obsolete flag when message is commented out, seems like best course of action.

Yes, that's awesome 👍 

> here is also support for msgctxt which denotes context. Based on reference doc, when the context is present, you can have a translation with the same msgid but with different context. That sounds like very useful feature in cases you are using messages as keys and sometimes it might be unclear in which context that message has been used. I have verified it with Loco service and it indeed works, some other will probably support it too. I will try to implement it from the side of format and perhaps there could a special macro that would add context to the message.

I wouldn't implement context, at least not right now. So far I saw only one or two uses and the workaround isn't difficult. On the other hand, library expects ID to be unique, so we either need to do massive refactoring and figure out some hack.

> One more thing... Currently, the description is considered to be a one-liner, but Gettext supports multiple #. lines for comments (official term - extracted comment). The pofile will parse all #. correctly and give out array of string. I don't like the idea of joining it together and then separating it again. I am thinking of keeping description prop to be a first comment and also to have comments with the rest of those. What do you think?

It's hard to say how external tools will handle `#.`... Lingui will always add/replace description of message, so if we assume that the first `#.` is description extracted from source then we're fine until external tool changes order of messages. This is "extracted comment", so the multiline description should be extracted from source by lingui, right? External tools shouldn't modify this description.

**Comment by FredyC** [comment]
> This is "extracted comment", so the multiline description should be extracted from source by lingui, right? External tools shouldn't modify this description.

I understand it the same way. However, I can imagine that a developer might want to add additional comments to a translation manually before sending it to a translator. In that case, it would make sense to keep additional comments there. It would probably need to be documented to use `#.` for additional comments and add it after the first line of description.

> I wouldn't implement context, at least not right now.

Yea, I've edited comment above probably after you read it :)

**Comment by tricoder42** [comment]
I would ignore it and raise a warning. Lingui CLI will never generate array or `msgstr`s, so it can only happen when someone imports a PO file with unsupported message format.

**Comment by tricoder42** [comment]
> I understand it the same way. However, I can imagine that a developer might want to add additional comments to a translation manually before sending it to a translator. In that case, it would make sense to keep those in there. It would probably need to be documented to use #. for additional comments and add it after the first line of description. Ideally, if this could be handled on source code level, to have multiple comments.

This is gonna be tricky. In such case the developer would edit locale-specific catalog, probably only in one locale, but right now the same description is used for all locales...

**Comment by FredyC** [comment]
> This is gonna be tricky. In such case, the developer would edit locale-specific catalog, probably only in one locale, but right now the same description is used for all locales...

You are right, it shouldn't be supported. It should show a warning that for any extract comments out of source code the pure `# comment` should be used. And I will actually transform any extra comments to translator comments. It would be just another property `comments` so it doesn't break anything.

> I would ignore it and raise a warning. Lingui CLI will never generate array or msgstrs, so it can only happen when someone imports a PO file with unsupported message format.

👍

**Comment by tricoder42** [comment]
FYI, at the moment the `description` property isn't used at all. It's just for the next major release, when description will be added to macros.

**Comment by FredyC** [comment]
Yea, that would be a reason why I was unable to find its use :) Have you actually thought about multiline descriptions? Because those should be probably split into multiple `#.` lines. If you plan something like that, I would prepare it right away to have it as an array.

**Comment by tricoder42** [comment]
The description can be any string. So, unless we prevent it, they can do something like this:

```js
t({ 
  id: "msg.id",
  description: `
    Message description
    (This is the second line)
  `
})
```

Even if we prevent it, they still can do `Message description\n(This is the second line)`

I'm very sceptical than anybody will actually use it, so it's probaly just *nice to have*.

**Comment by FredyC** [comment]
> Alright, I see now. The #~ seems to be an actual comment in PO file (see example). Based on this the module decides that particular translation is obsolete. I will set obsolete flag when message is commented out, seems like best course of action.

```
```

This probably needs a second thought. The `pofile` will parse `#~`, remove that notation from its internal structure and set `obsolete` property which is a different one from `obsolete` flag we use. When I see that prop I'll set our flag instead. When serializing back only the flag will be there and uncommented. Problem with this approach is, that flag is not official and won't be considered by translator tools.

Commented out messages are recognized correctly at least by POEdit and completely ignored while our flag is not really considered and the translator would spend time translating something that won't be used. Unless of course developer would run `clean` and remove it there.

**Comment by tricoder42** [comment]
Released in [v2.6.0](https://github.com/lingui/js-lingui/releases/tag/v2.6.0)
